File: dotnet/src/Envdoctor/Detectors/UnusedDetector.cs
using Envdoctor.Models;
using static Envdoctor.Detectors.DetectorHelpers;

namespace Envdoctor.Detectors;

/// Unused: a variable defined in an environment file that is never referenced
/// anywhere. `.env.example` contents are documentation and are excluded here.
public sealed class UnusedDetector : IDetector
{
    public string Id => "unused";
    public string Name => "unused";
    public string Description =>
        "Defined in an environment file but never referenced in source, docker-compose, GitHub Actions, or Kubernetes manifests.";

    public List<Finding> Detect(IndexedModel index)
    {
        var findings = new List<Finding>();
        var used = new HashSet<string>(index.Usages.Keys);
        used.UnionWith(index.ComposeDefinitions.Keys);
        used.UnionWith(index.ActionDefinitions.Keys);
        used.UnionWith(index.K8sDefinitions.Keys);

        // Iterate in a stable file/line order so output is deterministic and
        // matches the reference CLI.
        var entries = index.EnvDefinitions
            .OrderBy(kv => DefSortKey(kv.Value), Comparer<(string Path, int Line)>.Create(CompareKeys))
            .ThenBy(kv => kv.Key, StringComparer.Ordinal);

        var seen = new HashSet<string>();

        foreach (var (name, defs) in entries)
        {
            if (!seen.Add(name))
                continue;
            if (used.Contains(name))
                continue;
            findings.Add(MakeFinding(
                "unused",
                Severity.Warning,
                name,
                "defined but never referenced",
                defs.Select(d => d.Origin.Clone()).ToList()));
        }

        return findings;
    }
}

File: swift/Sources/EnvdoctorKit/Detectors/UnusedDetector.swift
import Foundation

/// Unused: a variable defined in an environment file that is never referenced
/// anywhere. `.env.example` contents are documentation and are excluded.
public struct UnusedDetector: Detector {
    public let id = "unused"
    public let name = "unused"
    public let description = "Defined in an environment file but never referenced in source, docker-compose, GitHub Actions, or Kubernetes manifests."

    public init() {}

    public func detect(_ index: IndexedModel) -> [Finding] {
        var findings: [Finding] = []
        var used = Set(index.usages.map(\.0))
        // A variable that is re-defined in compose/actions/k8s is, by definition, used.
        used.formUnion(index.composeDefinitions.map(\.0))
        used.formUnion(index.actionDefinitions.map(\.0))
        used.formUnion(index.k8sDefinitions.map(\.0))

        var seen = Set<String>()
        for (name, defs) in index.envDefinitions {
            if seen.contains(name) { continue }
            seen.insert(name)
            if used.contains(name) { continue }
            findings.append(makeFinding(
                "unused", .warning, name,
                "defined but never referenced",
                defs.map(\.origin)
            ))
        }
        return findings
    }
}
